README.md
# SharedStreets GeoJSON

[![npm version](https://badge.fury.io/js/sharedstreets-geojson.svg)](https://badge.fury.io/js/sharedstreets-geojson)
[![Build Status](https://travis-ci.org/sharedstreets/sharedstreets-geojson.svg?branch=master)](https://travis-ci.org/sharedstreets/sharedstreets-geojson)

Converts SharedStreets JSON to GeoJSON.

## Install

**In Node.js**

```bash
$ yarn add sharedstreets-geojson
```

**CommonJS**

```js
const sharedstreetsGeoJSON = require('sharedstreets-geojson');
```

**Typescript**

```js
import * as sharedstreetsGeoJSON from 'sharedstreets-geojson';
```

## In Browser

For a full list of web examples, check out [SharedStreets examples](https://github.com/sharedstreets/sharedstreets-examples).

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [geometry](#geometry)
-   [intersection](#intersection)

### geometry

Geometry

**Parameters**

-   `geom` **SharedStreetsGeometry** JSON SharedStreetsGeometry

**Examples**

```javascript
const line = [[110, 45], [115, 50], [120, 55]];
const geom = sharedstreets.geometry(line);
const geojson = sharedstreetsGeoJSON.geometry(geom)));
geojson // => Feature<LineString, SharedStreetsGeometry>
```

Returns **Feature&lt;LineString, SharedStreetsGeometry>** GeoJSON LineString Feature SharedSteetsGeometry

### intersection

Intersection

**Parameters**

-   `intersect` **SharedStreetsIntersection** JSON SharedStreetsIntersection

**Examples**

```javascript
const pt = [110, 45];
const intersect = sharedstreets.intersection(pt);
const geojson = sharedstreetsGeoJSON.intersection(intersect)));
geojson // => Feature<Point, SharedStreetsIntersection>
```

Returns **Feature&lt;Point, SharedStreetsIntersection>** GeoJSON Point Feature SharedSteetsIntersection


CHANGELOG.md
# Changelog

## 2018-03-08 - 0.1.0

- Implement `reference()` method
- Implement `intersection()` method
- Implement `geometry()` method
- Initialize `sharedstreets-geojson` (will be porting code from [`sharedstreets-python`](https://github.com/sharedstreets/sharedstreets-python))


test.ts
import { featureCollection } from "@turf/helpers";
import * as fs from "fs";
import * as glob from "glob";
import load from "load-json-file";
import * as path from "path";
import * as sharedstreetsPbf from "sharedstreets-pbf";
import { SharedStreetsGeometry, SharedStreetsIntersection, SharedStreetsReference } from "sharedstreets-types";
import test from "tape";
import write from "write-json-file";
import * as sharedstreetsGeoJSON from "./";

// Convert PBF to JSON if does not exist
glob.sync(path.join(__dirname, "test", "in", "*.pbf")).forEach((filepath) => {
  const { name } = path.parse(filepath);
  const ext = path.parse(name).ext.replace("\.", "");
  const filepathjson = filepath.replace(".pbf", ".json");
  if (!fs.existsSync(filepathjson)) {
    const buffer = fs.readFileSync(filepath);
    const parser: any = sharedstreetsPbf;
    const data = parser[ext](buffer);
    write.sync(filepathjson, data);
  }
});

test("sharedstreets-geojson -- geometry", (t) => {
  glob.sync(path.join(__dirname, "test", "in", "*.geometry.json")).forEach((filepath) => {
    // Load JSON Data
    const { name } = path.parse(filepath);
    const data: SharedStreetsGeometry[] = load.sync(filepath);

    // Convert JSON to GeoJSON
    const geojson = featureCollection(data.map((item) => sharedstreetsGeoJSON.geometry(item)));

    // Save GeoJSON
    const outpath = filepath.replace(path.join("test", "in"), path.join("test", "out")).replace(".json", ".geojson");
    if (process.env.REGEN) { write.sync(outpath, geojson); }
    t.deepEqual(geojson, load.sync(outpath));
  });
  t.end();
});

test("sharedstreets-geojson -- intersection", (t) => {
  glob.sync(path.join(__dirname, "test", "in", "*.intersection.json")).forEach((filepath) => {
    // Load JSON Data
    const { name } = path.parse(filepath);
    const data: SharedStreetsIntersection[] = load.sync(filepath);

    // Convert JSON to GeoJSON
    const geojson = featureCollection(data.map((item) => sharedstreetsGeoJSON.intersection(item)));

    // Save GeoJSON
    const outpath = filepath.replace(path.join("test", "in"), path.join("test", "out")).replace(".json", ".geojson");
    if (process.env.REGEN) { write.sync(outpath, geojson); }
    t.deepEqual(geojson, load.sync(outpath));
  });
  t.end();
});

test("sharedstreets-geojson -- reference", (t) => {
  glob.sync(path.join(__dirname, "test", "in", "*.reference.json")).forEach((filepath) => {
    // Load JSON Data
    const { name } = path.parse(filepath);
    const data: SharedStreetsReference[] = load.sync(filepath);

    // Convert JSON to JSON
    const references = data.map((item) => sharedstreetsGeoJSON.reference(item));

    // Save JSON
    const outpath = filepath.replace(path.join("test", "in"), path.join("test", "out"));
    if (process.env.REGEN) { write.sync(outpath, references); }
    t.deepEqual(references, load.sync(outpath));
  });
  t.end();
});


index.ts
import { Feature, featureCollection, FeatureCollection, lineString, LineString, point, Point } from "@turf/helpers";
import { lonlatsToCoords } from "sharedstreets";
import { SharedStreetsGeometry, SharedStreetsIntersection, SharedStreetsReference } from "sharedstreets-types";

interface SharedStreetsGeometryGeoJSON extends Feature<LineString, SharedStreetsGeometry> {
  role: "SharedStreets:Geometry";
}

interface SharedStreetsIntersectionGeoJSON extends Feature<Point, SharedStreetsIntersection> {
  role: "SharedStreets:Intersection";
}

interface SharedStreetsReferenceGeoJSON extends SharedStreetsReference {
  role: "SharedStreets:Reference";
}

interface SharedStreetsGeoJSON extends FeatureCollection {
  references: SharedStreetsReferenceGeoJSON[];
}

/**
 * Geometry
 *
 * @param {SharedStreetsGeometry} geom JSON SharedStreetsGeometry
 * @returns {Feature<LineString, SharedStreetsGeometry>} GeoJSON LineString Feature SharedSteetsGeometry
 * @example
 * const line = [[110, 45], [115, 50], [120, 55]];
 * const geom = sharedstreets.geometry(line);
 * const geojson = sharedstreetsGeoJSON.geometry(geom)));
 * geojson // => Feature<LineString, SharedStreetsGeometry>
 */
export function geometry(geom: SharedStreetsGeometry): SharedStreetsGeometryGeoJSON {
  const coords = lonlatsToCoords(geom.lonlats);
  const properties = geom;

  // Do not include (lonlats) to GeoJSON properties
  if (properties.lonlats) { delete properties.lonlats; }

  // Add GeoJSON Foreign Member (role)
  const line: any = lineString(coords, properties);
  line.role = "SharedStreets:Geometry";
  return line;
}

/**
 * Intersection
 *
 * @param {SharedStreetsIntersection} intersect JSON SharedStreetsIntersection
 * @returns {Feature<Point, SharedStreetsIntersection>} GeoJSON Point Feature SharedSteetsIntersection
 * @example
 * const pt = [110, 45];
 * const intersect = sharedstreets.intersection(pt);
 * const geojson = sharedstreetsGeoJSON.intersection(intersect)));
 * geojson // => Feature<Point, SharedStreetsIntersection>
 */
export function intersection(intersect: SharedStreetsIntersection): SharedStreetsIntersectionGeoJSON {
  const coords = [intersect.lon, intersect.lat];
  const properties = intersect;

  // Do not include (lon, lat) to GeoJSON properties
  if (properties.lon) { delete properties.lon; }
  if (properties.lat) { delete properties.lat; }

  // Add GeoJSON Foreign Member (role)
  const pt: any = point(coords, properties);
  pt.role = "SharedStreets:Intersection";
  return pt;
}

/**
 * Reference
 *
 * @param {SharedStreetsReference} intersect JSON SharedStreetsReference
 * @returns {SharedStreetsReference} SharedSteetsIntersection with Role
 */
export function reference(ref: SharedStreetsReference): SharedStreetsReferenceGeoJSON {
  const data: any = ref;
  data.role = "SharedStreets:Reference";
  return data;
}

/**
 * GeoJSON
 *
 * @param {Array<SharedStreetsGeometry>} geometries An Array of SharedStreetsGeometry
 * @param {Array<SharedStreetsIntersection>} intersections An Array of SharedStreetsIntersection
 * @param {Array<SharedStreetsReference>} references An Array of SharedStreetsReference
 * @returns {FeatureCollection} SharedStreets GeoJSON
 */
export function geojson(
  geometries: SharedStreetsGeometry[],
  intersections: SharedStreetsIntersection[],
  references: SharedStreetsReference[],
): SharedStreetsGeoJSON {
  const data: any = featureCollection([]);

  geometries.forEach((item) => data.features.push(geometry(item)));
  intersections.forEach((item) => data.features.push(intersection(item)));
  references.forEach((item) => data.references.push(reference(item)));
  return data;
}
